Relocate SSNv2 range objects for new CA instances

edewata opened the pull request:
`pkispawn` has been modified to create `ou=ranges` subtree for SSNv1 and optionally `ou=ranges_v2` subtree for SSNv2 if it's enabled for new CA instances. The `pki-server <subsystem>-db-init` and `<subsystem>-range-update` commands have been updated to use the proper subtree to store the range objects. Hard-coded subtrees in the `create.ldif` have been removed.

Similar changes are made to KRA as well, but since there are no tests for KRA with SSNv2 it's not officially supported yet.

The test scripts have been updated to use `ou=ranges_v2` subtree for SSNv2 range objects.

edewata commented:
@fmarco76 For SSNv2 we could use it like this: https://github.com/edewata/pki/commit/4a85681a43a866e699ef7c150b80ec91b9331e3b

fmarco76 commented:
> @fmarco76 For SSNv2 we could use it like this: [edewata@4a85681](https://github.com/edewata/pki/commit/4a85681a43a866e699ef7c150b80ec91b9331e3b)
Since the DS was configurable from the start it is OK to add the possibility to configure. To move the `nextRange` I am not sure yet.

edewata commented:
@fmarco76 Thanks for the comments! I'll update this PR after PR #4886. I think we need to change the location of `nextRange` for SSNv2 so that if there's a problem with the migration the `nextRange` for SSNv1 is still intact.

fmarco76 commented:
>  For SSNv2 we could use it like this: [edewata@4a85681](https://github.com/edewata/pki/commit/4a85681a43a866e699ef7c150b80ec91b9331e3b)

With the new commit PR #4887 the current ranges tree is not modified so it is alwais possible to get the `nextRange` value for the legacy generator therefore I am wondering if it could be enough a procedure to revert the nextRange value. May be with the command:

```
# pki-server ca-id-generator-update --type legacy --revert cert
```
The option `--revert` means: do not do migration but just revert the value in `CS.cfg` and in nextRange using the range values and eventually some of the `dbs.*` values.

edewata commented:
As mentioned in PR #4887 we are still modifying the same `nextRange` attribute used by SSNv1 which is replicated to all DS instances and it might create potential issues. I'd prefer we move that into the new `ou=ranges_v2` subtree so we have a cleaner separation between SSNv1 and SSNv2. Once this is done, if we provide a revert functionality it will be just restoring the SSNv1 params in `CS.cfg`. However, if new certs have been issued with SSNv2 things could be more complicated. I'm not sure how SSNv1 code will handle that scenario. Worst case the new certs will need to be removed/revoked as well. See also the Recovery section for RSNv3: https://github.com/dogtagpki/pki/wiki/Configuring-CA-with-Random-Serial-Numbers-v3

fmarco76 commented:
> However, if new certs have been issued with SSNv2 things could be more complicated. I'm not sure how SSNv1 code will handle that scenario. Worst case the new certs will need to be removed/revoked as well. See also the Recovery section for RSNv3: https://github.com/dogtagpki/pki/wiki/Configuring-CA-with-Random-Serial-Numbers-v3

There should be no problem to move from ssnv1 to ssnv2 and going back. The certificate are serial and there is no overlap issue so the problem is to set the right value to avoid conflicts. In the worst case it is possible to read last serial released and start new range from that point. You could have a big gap but not issues.

To move the nextRange in a different location additional configuration parameters could be needed. It is weird that looking at the repository object the nextRange is not used and maintain a wrong  value while the new repository for nextRange actually it is not a repository. Therefore, I am also looking at the possibility to store the nextRange in a different place (e.g. the description field is not used, could it be used?)

edewata commented:
> > However, if new certs have been issued with SSNv2 things could be more complicated. I'm not sure how SSNv1 code will handle that scenario. Worst case the new certs will need to be removed/revoked as well. See also the Recovery section for RSNv3: https://github.com/dogtagpki/pki/wiki/Configuring-CA-with-Random-Serial-Numbers-v3
> 
> There should be no problem to move from ssnv1 to ssnv2 and going back. Certificates are serial and there is no overlap issue so the problem is to find the right values to avoid conflicts. In the worst case it is possible to read last serial released and start new range from that point. You could have a big gap but not issues.

Yes, it might work just fine, but since we do not have a test for this, we don't really know whether there's any unforeseen issues. It's better to minimize the risk as much as possible, especially since we will not be able to assist directly during the actual migration of production environment.
 
> To move the nextRange in a different location additional configuration parameters could be needed. It is weird that looking at the repository object the nextRange is not used and maintain a wrong value while the new repository for nextRange actually it is not a repository. Therefore, I am also looking at the possibility to store the nextRange in a different place (e.g. the description field is not used, could it be used?)

In https://github.com/edewata/pki/commit/4a85681a43a866e699ef7c150b80ec91b9331e3b I was actually trying to use `repository` object class for the ranges subtree so we can store the `nextRange` there. Let me update this PR to include that change.

edewata commented on base/server/python/pki/server/deployment/__init__.py:
```diff
@@ -1207,6 +1207,10 @@ def configure_ca(self, subsystem):
+            request_dn = self.mdict.get('pki_request_number_range_dn')
+            if request_dn:
+                subsystem.set_config('dbs.requestRangeDN', request_dn)
+
```
I decided not to add the `pkispawn` param for now due to complexity, so the subtree is always `ou=ranges` for SSNv1 and `ou=ranges_v2` for SSNv2.

edewata commented on base/server/python/pki/server/deployment/__init__.py:
```diff
@@ -1245,6 +1249,10 @@ def configure_ca(self, subsystem):
+            serial_dn = self.mdict.get('pki_serial_number_range_dn')
+            if serial_dn:
+                subsystem.set_config('dbs.serialRangeDN', serial_dn)
+
```
Please see above.

edewata commented on base/server/src/main/java/org/dogtagpki/server/cli/SubsystemDBInitCLI.java:
```diff
@@ -105,6 +109,21 @@ public void execute(CommandLine cmd) throws Exception {
                 ldapConfigurator.createContainers(subsystem);
+
+                String requestRangeRDN = dbConfig.getRequestRangeDN();
+                if (!StringUtils.isEmpty(requestRangeRDN)) {
```
In the updated PR the param is guaranteed to exist in CA and KRA so it won't fail. The param will not exist in other subsystems so the code won't create the subtree.

edewata commented:
@fmarco76 Sorry for the delay. Please see the updated PR. I've changed the code to create SSNv2 range objects in `ou=ranges_v2` for new instances to match migrated instances. I dropped the `pkispawn` params for now and deferred moving the `nextRange` to a separate PR due to complexity. There are some failures in the CI but I think they are caused by recent changes in Fedora.

fmarco76 reviewed (approved):
LGTM

edewata commented:
@fmarco76 Thanks!
